src/main/java/seedu/address/model/VersionedAddresSoc.java

```diff
@@ -0,0 +1,64 @@
+package seedu.address.model;
+
+import java.util.ArrayList;
+
+/**
+ * Keeps track of changes to AddresSoc and implements
+ * the Undo and Redo functions.
+ */
+public class VersionedAddresSoc {
+    private final ArrayList<ReadOnlyAddressBook> addressBookStateList;
+    private final AddressBook initialAddressBook;
+    private int currentStatePointer;
+
+
+    /**
+     * Initializes VersionedAddresSoc with the initial AddressBook.
+     */
+    public VersionedAddresSoc(AddressBook initialAddressBook) {
+        this.initialAddressBook = initialAddressBook;
+        ReadOnlyAddressBook initialState = new AddressBook(initialAddressBook);
+        addressBookStateList = new ArrayList<ReadOnlyAddressBook>();
+        addressBookStateList.add(initialState);
+        currentStatePointer = 0;
+    }
+
+    /**
+     * Commits current AddressBook to the StateList as a State.
+     * @param currentAddressBook to be saved in the StateList
+     */
+    public void commit(ReadOnlyAddressBook currentAddressBook) {
+        ReadOnlyAddressBook currentState = new AddressBook(currentAddressBook);
+        if (addressBookStateList.size() > currentStatePointer + 1) {
+            addressBookStateList.subList(currentStatePointer + 1, addressBookStateList.size()).clear();
+        }
+        currentStatePointer++;
+        addressBookStateList.add(currentState);
+    }
+
+    /**
+     * Decrement pointer and resets the AddressBook to its previous state.
+     */
+    public void undo() {
+        currentStatePointer--;
+        ReadOnlyAddressBook newState = addressBookStateList.get(currentStatePointer);
```

Comment:
There could be an index out of bounds exception if someone does not check `canUndo()` before calling this method. Maybe it might be good to add an assertion here to check it and document the precondition in the javadoc comment.